### PR TITLE
Add option for quiet app init. 

### LINKED ISSUE #24264: Allow "non-competitive" initialization of sub-apps
## Reason
Cardinal runs NekRS simulations within MOOSE. If we have > 1 NekRS simulation in a given run (such as when multiple Nek sub-apps are provided with the `MultiApp/positions` parameter, or when running a StochasticTools simulation with `num_rows` > 1), `MultiApp::createApps` will try to create all N NekRS sub-apps concurrently.

However, NekRS is not a MOOSE-based application, and has some special requirements on running > 1 NekRS case concurrently. NekRS does JIT compilation of GPU kernels, plus a bunch of other stuff, and writes temporary files to a `.cache/` directory in the same folder as your input files.

```
cardinal-opt -i nek.i

# first thing this does is JIT compile and write files into a .cache/ directory
```

This JIT compile MUST happen in a "quiet" environment. If a `MultiApp` is trying to spawn 5 different NekRS sub-apps, they will clobber each other and overwrite the contents of the `.cache/`, causing catastrophic failures. 

We are integrating NekRS with the Stochastic Tools module. If we want to run a stochastic simulation with 500 samples, this means that we either:

- Need to launch our job with 500 MPI ranks *and* set `min_procs_per_app = 500` so that only a single Nek sub-app is created. This is not going to be feasible for most scenarios, because the number of stochastic samples is typically quite high, imposing artificial constraints on the number of processes in a parallel job (and tangible impacts on long queue times). This also could be a bad idea for performance, if each individual Nek simulation is small, we don't want to artificially force them to run with loads of processes.
- OR, add some logic that will "pause" the activity of all non-rank-0 MPI processes while the very first Nek sub-app is JIT compiled, and then resume parallel instantiation.

## Design
Add an option to `MultiApp` which changes how the sub-apps are created, so that the first app on rank 0 is constructed "in quiet" while everybody else waits. Only after that, let all the other ranks resume initialization of their apps. 

I think this is the simplest way to accomplish this, but am open to suggestions.

## Impact
New feature. Necessary for NEAMS milestone.

### PR DESCRIPTION
**Pull request by aprilnovak**
I tested this in Cardinal, and find this to be adequate for our needs.  

Closes #24264

### CODE REVIEW

**Comment by aprilnovak** [comment]
I think the Okami failures are unrelated.

**Comment by aprilnovak** [comment]
Actually, I discovered that this only works if I run with less ranks than # sub-apps. Investigating further.

**Review by lindsayad — CHANGES_REQUESTED** [review]
Can we test this?

**Comment by aprilnovak** [comment]
I am not sure how... we'd need some type of Action/App which like, proves there's a race condition when > 1 App is constructing itself at the same time. 

I do have tests in Cardinal, but I'm not sure of a simple way to test this just in MOOSE. Do you have any ideas?

**Comment by aprilnovak** [comment]
I could create a test which shows this has no effect for normal MOOSE Apps? (since they don't care if they construct at the same time)

**Comment by GiudGiud** [comment]
RunApp with expect_out and absent_out doesnt do the trick ?

**Comment by GiudGiud** [comment]
Note to self, this could really go into the [Application] block (when it exists)
#23741

**Comment by aprilnovak** [comment]
> RunApp with expect_out and absent_out doesnt do the trick ?

Can you elaborate more on what you mean? Are you saying that I should put some print statement inside of my new if-statement, and test whether that gets printed?

File: framework/src/multiapps/MultiApp.C [hunk 1]
```diff
@@ -216,7 +224,7 @@ MultiApp::validParams()
 
   params.addParamNamesToGroup("positions positions_file run_in_position output_in_position",
                               "Positions / transformations of the MultiApp frame of reference");
-  params.addParamNamesToGroup("min_procs_per_app max_procs_per_app", "Parallelism");
+  params.addParamNamesToGroup("min_procs_per_app max_procs_per_app quiet_app_init", "Parallelism");
```
Comment (GiudGiud) [inline framework/src/multiapps/MultiApp.C:227]: move it in the Advanced group maybe?

File: framework/src/multiapps/MultiApp.C [hunk 2]
```diff
@@ -120,6 +120,14 @@ MultiApp::validParams()
                                 1,
                                 "Minimum number of processors to give to each App in this "
                                 "MultiApp.  Useful for larger, distributed mesh solves.");
+  params.addParam<bool>(
+      "quiet_app_init",
```
Comment (GiudGiud) [inline framework/src/multiapps/MultiApp.C:124]: it s more "staggered" than quiet?

but even staggered would be 1 by 1 not 1 then n-1
suggestions? wish there was a "volunteer-naming-suggestions" group we could tag haha

File: test/tests/multiapps/centroid_multiapp/tests [hunk 3]
```diff
@@ -9,4 +9,14 @@
...
   [../]
+  [test_quiet]
+    type = Exodiff
+    input = centroid_multiapp.i
+    cli_args = 'MultiApps/sub/quiet_app_init=true'
+    min_parallel = 2
+    recover = false
+    exodiff = 'centroid_multiapp_out.e centroid_multiapp_out_sub1.e centroid_multiapp_out_sub3.e centroid_multiapp_out_sub5.e centroid_multiapp_out_sub7.e centroid_multiapp_out_sub0.e centroid_multiapp_out_sub2.e centroid_multiapp_out_sub4.e centroid_multiapp_out_sub6.e centroid_multiapp_out_sub8.e'
+
+    requirement = "The system shall optionally allow the first sub-app to fully initialize before launching the remaining N-1 apps in parallel."
```
Comment (GiudGiud) [inline test/tests/multiapps/centroid_multiapp/tests:20]:
```suggestion
    requirement = "The system shall optionally allow the first sub-app to fully initialize before initializing all the remaining child apps in parallel."
```

**Review by GiudGiud — COMMENTED** [review]
nevermind the other comment. This PR wasnt what I though

File: framework/src/multiapps/MultiApp.C (hunk 2 above)
Comment (aprilnovak) [reply framework/src/multiapps/MultiApp.C:124]: Yeah I was not sure of a good name here - "staggered" is perhaps closer to the intention... `barrier_until_first_app_initialized` is probably the most literal choice?

Comment (loganharbour) [reply framework/src/multiapps/MultiApp.C:123]: `wait_for_first_app`

Comment (GiudGiud) [reply framework/src/multiapps/MultiApp.C:123]: works
wait_for_first_app_init ?

Comment (loganharbour) [reply framework/src/multiapps/MultiApp.C:123]: Either one is fine with me. I think they're both a little more intuitive

Comment (aprilnovak) [reply framework/src/multiapps/MultiApp.C]: I went with `wait_for_first_app_init`, I agree those are better names.

File: framework/include/multiapps/MultiApp.h [hunk 4]
```diff
@@ -131,6 +131,12 @@ class MultiApp : public MooseObject,
    */
   void setupPositions();
 
+  /**
+   * Create the i-th local app
+   * @param[in] i local app index
+   */
+  virtual void createLocalApp(const unsigned int & i);
```
Comment (lindsayad) [inline framework/include/multiapps/MultiApp.h:138]:
```suggestion
  virtual void createLocalApp(const unsigned int i);
```
It's usually faster to copy builtins

File: framework/src/multiapps/MultiApp.C [hunk 5]
```diff
@@ -357,13 +365,35 @@ MultiApp::createApps()
                                 getParam<std::string>("library_name"),
                                 getParam<bool>("library_load_dependencies"));
 
+  bool rank_did_quiet_init = false;
+  unsigned int local_app;
```
Comment (lindsayad) [inline framework/src/multiapps/MultiApp.C:369]:
```suggestion
  unsigned int local_app = libMesh::invalid_uint;
```
I don't know if valgrind would warn about this (I've added that recipe), but it makes me a little worried

**Review by lindsayad — CHANGES_REQUESTED** [review]
Thank you so much for the test. Just these small requests (the valgrind one is the only reason I'm requesting changes as opposed to approving with comments)

**Review by lindsayad — CHANGES_REQUESTED** [review]
Thank you so much for the test. Just these small requests (the valgrind one is the only reason I'm requesting changes as opposed to approving with comments)

File: framework/include/multiapps/MultiApp.h (hunk 4 above)
Comment (GiudGiud) [reply framework/include/multiapps/MultiApp.h:138]: Roy had a good comment on reference vs value
https://github.com/idaholab/moose/pull/21904#discussion_r1172738030

**Review by lindsayad — APPROVED** [review]

**Pull request merged**